File: cpp-implementation/src/products/Books/NonFiction.cpp
#include "Product.ih"

// Updated constructor implementation for NonFiction books
NonFiction::NonFiction(int id, string name, double price, int volume,
                       string author, string publisher, string field, string supplier)
    : Books(id, name, price, volume, author, publisher, supplier),
      field(field) {}

// Calculates the discount depending on the current month and the packaged quantity
double NonFiction::calculateDiscount(int volume, int month) const {
    if (month == 12) {
        return price * volume * 0.25;
    }
    return 0;
}

// Displays all the information 
string NonFiction::displayProductInfo() const {
    ostringstream info;
    info << "ID: " << id << "\nNon-Fiction Book: " << name << "\nSupplier: " << getSupplier()
         << "\nAuthor: " << author << "\nPublisher: " << publisher 
         << "\nField: " << field << "\nPrice: €" << price 
         << "\nVolume: " << volume << " units\n\n";
    return info.str();
}

File: java-implementation/src/main/java/com/inventory/products/books/NonFiction.java
package com.inventory.products.books;

// Represents a concrete subclass of Books
public class NonFiction extends Books {
    private String field;

    // Constructs an instance with specified details with base class attributes included
    public NonFiction(int id, String name, double price, int volume, String author, String publisher, String field, String supplier) {
        super(id, name, price, volume, author, publisher, supplier);
        this.field = field;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 12) {
            return price * volume * 0.25;
        }
        return 0;
    }

    // Provides a string representation the details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nNon-Fiction Book: " + name + "\nAuthor: " + author
               + "\nPublisher: " + publisher + "\nField: " + field
               + "\nPrice: €" + price + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}
